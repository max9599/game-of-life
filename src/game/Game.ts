import Config from './Config'

export interface GameState {
    grid: number[]
    iteration: number
    paused: boolean
    config: Config
    start: any
    stop: any
    restart: any
}

export default class Game {

    config: Config = new Config()
    iteration: number = 0
    paused: boolean = false
    grid: number[] = []

    interval?: number
    update?: (game: GameState) => void

    constructor(config: Config) {
        this.config = config
    }

    initGame = () => {
        const size: number = this.config.size
        for (let i = 0, len = size * size; i < len; i++) {
            this.initXYCell(i)
        }
    }

    start = (reset?: boolean) => {
        if (reset) {
            this.iteration = 0
            this.grid = []
            this.config.fillInitialAliveCellsMap()
            this.initGame()
        }
        this.paused = false
        this.interval = setInterval(this.tick, this.config.intervalMs)
        this.tick()
        return this
    }

    onUpdate = (doAction: (state: GameState) => void) => {
        this.update = doAction
        return this
    }

    stop = () => {
        this.paused = true
        clearInterval(this.interval)
        this.doUpdate()
    }

    restart = () => {
        this.stop()
        this.start(true)
    }

    private tick = () => {
        this.iteration++
        const grid = []
        for (let i = 0, gridLen = this.grid.length; i < gridLen; i++) {
            const total = this.getTotalAliveNeighbours(i)
            grid[i] = total !== 2 ? (total === 3 ? 1 : 0) : this.grid[i]
        }
        this.grid = grid
        this.doUpdate()
        if (this.iteration >= this.config.iterations) {
            this.stop()
        }
    }

    doUpdate = () => {
        if (this.update) {
            this.update(this.getGameState())
        }
    }

    getGameState = () => ({
        paused: this.paused,
        grid: this.grid,
        iteration: this.iteration,
        config: this.config,
        start: this.start,
        stop: this.stop,
        restart: this.restart,
    })

    private getTotalAliveNeighbours = (pos: number): number => {
        const { size } = this.config
        return this.grid[pos - 1 - (size)] +
            this.grid[pos - (size)] +
            this.grid[pos + 1 - (size)] +
            this.grid[pos - 1] +
            this.grid[pos + 1] +
            this.grid[pos - 1 + (size)] +
            this.grid[pos + (size)] +
            this.grid[pos + 1 + (size)]
    }

    private initXYCell = (i: number) => {
        const { initialAliveCellsIndexes } = this.config
        const alive = initialAliveCellsIndexes.indexOf(i) >= 0
        this.grid[i] = alive ? 1 : 0
    }
}
