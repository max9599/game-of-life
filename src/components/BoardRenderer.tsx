import React, { useRef, useEffect } from 'react'
import styled from 'styled-components'
import { GameState } from '../game/Game'

export type BoardRendererProps = {
    game: GameState
}

const BoardStyled = styled.div`
    position: absolute;
    overflow: auto;
    max-height: 100vh;
    right: 0;
    top: 0;
    height: 100vh;
    width: calc(100vw - 250px);
    background: rgba(0,0,0,0.01);
    
    canvas {
        border: 1px solid rgba(0,0,0,.05);
        background: rgba(0,0,0,0.0ยง);
    }
`

const BoardRenderer: React.FC<BoardRendererProps> = ({ game }) => {
    const { grid, config } = game
    const { unit, size, cellRGBA } = config
    const canvasElement = useRef<HTMLCanvasElement>(null)
    const sizeInPixel = size * unit
    useEffect(() => {
        if (canvasElement.current) {
            const canvas = canvasElement.current
            const context = canvas.getContext('2d')
            if (context) {
                const imageData = context.createImageData(sizeInPixel, sizeInPixel)
                const { data } = imageData
                for (let i = 0, gridLen = size * size; i < gridLen; i++) {
                    if (grid[i]) {
                        const x = ~~(i % size) * unit
                        const y = ~~(i / size) * unit
                        for (let pixelIndex = 0, len = unit * unit; pixelIndex < len; pixelIndex++) {
                            const xx = ~~(pixelIndex % unit)
                            const yy = ~~(pixelIndex / unit)
                            const pData = ((x + xx) + (y + yy) * sizeInPixel) * 4
                            data[pData] = cellRGBA[0]
                            data[pData + 1] = cellRGBA[1]
                            data[pData + 2] = cellRGBA[2]
                            data[pData + 3] = cellRGBA[3]
                        }
                    }
                }
                context.putImageData(imageData, 0, 0)
            }
        }
    })

    return <BoardStyled data-testid="board-renderer">
        <canvas ref={canvasElement}
                width={sizeInPixel}
                height={sizeInPixel}/>
    </BoardStyled>
}

export default React.memo(BoardRenderer)
