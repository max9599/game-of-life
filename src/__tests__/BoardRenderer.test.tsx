import React from 'react'
import { render } from '@testing-library/react'
import 'jest-canvas-mock'

import BoardRenderer, { BoardRendererProps } from '../components/BoardRenderer'
import Game, { Config } from '../game'

const applyPattern = (config: Config) => config
    .addInitialAliveCell(52, 26)
    .addInitialAliveCell(52, 26)
    .addInitialAliveCell(52, 27)
    .addInitialAliveCell(51, 27)
    .addInitialAliveCell(52, 28)
    .addInitialAliveCell(53, 28)

const defaultConfig = () => new Config()
    .setSize(10)
    .setUnit(1)
    .setIntervalMs(100)
    .setCellRGBA([0, 55, 0, 255])
    .setIterations(10)

async function renderBoardRendererAndGetCanvas(props: Partial<BoardRendererProps> = {}, customConfig?: Config) {
    const game = new Game(customConfig || defaultConfig())
    const defaultProps: BoardRendererProps = { game }
    const overrideGame = props.game ? { game: { ...props.game, ...defaultProps.game } } : defaultProps
    const component = render(<BoardRenderer {...overrideGame} />)
    const { firstChild } = await component.findByTestId('board-renderer')
    return firstChild as HTMLCanvasElement
}

describe('<BoardRenderer />', () => {
    test('should validate not in range X / Y and throw error', async () => {
        const config = defaultConfig()
        expect(() => applyPattern(config.setSize(10))).toThrowError('X / Y is not in specified range')
    })
    test('should render board with initial state of game', async () => {
        const config = applyPattern(defaultConfig().setSize(100))
        const canvas: HTMLCanvasElement = await renderBoardRendererAndGetCanvas({}, config)
        expect(canvas.width).toBe(100)
        expect(canvas.height).toBe(100)
        const ctx = canvas.getContext('2d')
        expect(ctx).not.toBeNull()
        const events = ctx!!.__getEvents()
        expect(events.length).toBe(2)
        expect(events[0]).toEqual(expect.objectContaining({ type: 'createImageData' }))
        expect(events[1]).toEqual(expect.objectContaining({ type: 'putImageData' }))
        expect(events).toMatchSnapshot()
    })
    test('should render board with size depending on unit', async () => {
        const config = applyPattern(defaultConfig().setSize(100).setUnit(5))
        const canvas: HTMLCanvasElement = await renderBoardRendererAndGetCanvas({}, config)
        expect(canvas.width).toBe(500)
        expect(canvas.height).toBe(500)
    })
})
