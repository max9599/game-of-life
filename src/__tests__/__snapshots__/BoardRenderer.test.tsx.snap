// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BoardRenderer /> should render board with initial state of game 1`] = `
Array [
  Object {
    "props": Object {
      "height": 100,
      "width": 100,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "createImageData",
  },
  Object {
    "props": Object {
      "dirtyHeight": NaN,
      "dirtyWidth": NaN,
      "dirtyX": NaN,
      "dirtyY": NaN,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "putImageData",
  },
]
`;
